src/language-server/language-server.ts

```diff
@@ -0,0 +1,239 @@
+  // Make the text document manager listen on the connection
+  // for open, change and close text document events
+  documents.listen(connection)
+
+  connection.onCodeAction((params) => {
+    const document = documents.get(params.textDocument.uri)
+    if (!document) {
+      return []
+    }
+
+    const codeActions: CodeAction[] = []
+
+    return codeActions
+  })
+
+  connection.listen()
+}
+
+async function parse(connection: Connection, documents: TextDocuments<TextDocument>, path: string): Promise<Map<string, Diagnostic[]>> {
+  try {
+    connection.console.log(`Parsing ${path}`)
+
+    // TODO: test new file saved, then ref by an existing file
+    const files = processInputPaths({ paths: [path], outDir: 'tests/virtual-file/out' })
+
+    // To support unsaved files, we need to replace the file content with the content of the document
+    const processedFiles = files.map((file) => {
+      const fileUri = URI.file(file.sourceFile).toString()
+      const document = documents.get(fileUri)
+
+      if (document) {
+        return {
+          sourceFile: file.sourceFile,
+          outDir: 'tests/virtual-file/out',
+          fileContents: document.getText(),
+        }
+      }
+      return file
+    })
+
+    const logEvents = await compileFiles(processedFiles)
+
+    // Filter for errors and warnings
+    const filteredEvents = logEvents.filter((e) => e.level === LogLevel.Error || e.level === LogLevel.Warning)
+
+    // Group diagnostics by file URI
+    const diagnosticsMap = new Map<string, Diagnostic[]>()
+
+    for (const event of filteredEvents) {
+      if (!event.sourceLocation || !event.sourceLocation.file) continue
+
+      const fileUri = URI.file(event.sourceLocation.file).toString()
+
+      const diagnostic: Diagnostic = {
+        source: 'Puya TS',
+        severity: event.level === LogLevel.Error ? DiagnosticSeverity.Error : DiagnosticSeverity.Warning,
+        range: {
+          start: {
+            line: event.sourceLocation.line - 1,
+            character: event.sourceLocation.column,
+          },
+          end: {
+            line: event.sourceLocation.endLine - 1,
+            character: event.sourceLocation.endColumn,
+          },
+        },
+        message: event.message,
+      }
+
+      if (!diagnosticsMap.has(fileUri)) {
+        diagnosticsMap.set(fileUri, [])
+      }
+
+      diagnosticsMap.get(fileUri)!.push(diagnostic)
+    }
+
+    return diagnosticsMap
+  } catch (error) {
+    connection.console.error(`Failed to parse: ${JSON.stringify(error)}`)
+    return new Map()
+  }
+}
+
+async function compileFiles(files: AlgoFile[]) {
+  const logCtx = LoggingContext.create()
+  await logCtx.run(async () => {
+    await compile(
+      new CompileOptions({
+        filePaths: files,
+        dryRun: false,
+      }),
+    )
+  })
+  return logCtx.logEvents
+}
+
+function debounce<T extends (...args: any[]) => Promise<any>>(func: T, wait: number): (...args: Parameters<T>) => Promise<ReturnType<T>> {
```

Comment:
This feels like a potentially simpler implementation
https://jsdev.space/snippets/debounce-ts/